File: Solution with Bitset for recording ancestors.go

package main

import (
    "container/list"
    "fmt"
    "math/big"
)

var numberOfNodes int

func getAncestors(totalNodes int, edges [][]int) [][]int {
    numberOfNodes = totalNodes
    incomingEdgesPerNode := createArrayIncomingEdgesPerNode(edges)
    directedAcyclicGraph := createDirectedAcyclicGraph(edges)

    return findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph, incomingEdgesPerNode)
}

func findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph [][]int, incomingEdgesPerNode []int) [][]int {
    queue := createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode)
    ancestors := make([]big.Int, numberOfNodes)
    for nodeID := 0; nodeID < numberOfNodes; nodeID++ {
        ancestors[nodeID] = big.Int{}
    }

    for queue.Len() > 0 {
        current := queue.Front().Value.(int)
        queue.Remove(queue.Front())

        for _, next := range directedAcyclicGraph[current] {
            updateAncestors(current, next, ancestors)
            incomingEdgesPerNode[next]--
            if incomingEdgesPerNode[next] == 0 {
                queue.PushBack(next)
            }
        }
    }
    return createListAllAncestorsPerNode(ancestors)
}

func createArrayIncomingEdgesPerNode(edges [][]int) []int {
    incomingEdgesPerNode := make([]int, numberOfNodes)
    for _, edge := range edges {
        to := edge[1]
        incomingEdgesPerNode[to]++
    }
    return incomingEdgesPerNode
}

func createDirectedAcyclicGraph(edges [][]int) [][]int {
    directedAcyclicGraph := make([][]int, numberOfNodes)
    for nodeID := 0; nodeID < numberOfNodes; nodeID++ {
        directedAcyclicGraph = append(directedAcyclicGraph, []int{})
    }

    for _, edge := range edges {
        from := edge[0]
        to := edge[1]
        directedAcyclicGraph[from] = append(directedAcyclicGraph[from], to)
    }
    return directedAcyclicGraph
}

func createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode []int) *list.List {
    queue := list.New()
    for nodeID := 0; nodeID < numberOfNodes; nodeID++ {
        if incomingEdgesPerNode[nodeID] == 0 {
            queue.PushBack(nodeID)
        }
    }
    return queue
}

func createListAllAncestorsPerNode(ancestors []big.Int) [][]int {
    allAncestorsPerNode := make([][]int, numberOfNodes)

    for nodeID := 0; nodeID < numberOfNodes; nodeID++ {
        ancestorsForCurrentNode := []int{}

        for candidateAncestorID := 0; candidateAncestorID < numberOfNodes; candidateAncestorID++ {
            if ancestors[nodeID].Bit(candidateAncestorID) == 1 {
                ancestorsForCurrentNode = append(ancestorsForCurrentNode, candidateAncestorID)
            }
        }
        allAncestorsPerNode[nodeID] = ancestorsForCurrentNode
    }
    return allAncestorsPerNode
}

func updateAncestors(current int, next int, ancestors []big.Int) {
    ancestors[next].Or(&ancestors[next], &ancestors[current])
    ancestors[next].SetBit(&ancestors[next], current, 1)
}

File: Solution with Bitset for recording ancestors.cpp

#include <span>
#include <queue>
#include <vector>
#include <bitset>
using namespace std;

class Solution {

    static constexpr array<int,2> RANGE_OF_NODES = {1,1000};
    int numberOfNodes{};

public:
    vector<vector<int>> getAncestors(int numberOfNodes, const vector<vector<int>>& edges) {
        this->numberOfNodes = numberOfNodes;
        vector<int> incomingEdgesPerNode = createArrayIncomingEdgesPerNode(edges);
        vector<vector<int>> directedAcyclicGraph = createDirectedAcyclicGraph(edges);

        return findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph, incomingEdgesPerNode);
    }

private:
    vector<vector<int>> findAncestorsPerNodeWithTopologicalSort(span<const vector<int>> directedAcyclicGraph, span<int> incomingEdgesPerNode) const {
        queue<int> queue = createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode);
        vector<bitset<RANGE_OF_NODES[1] + 1>> ancestors(numberOfNodes);

        while (!queue.empty()) {
            int current = queue.front();
            queue.pop();

            for (const auto& next : directedAcyclicGraph[current]) {
                updateAncestors(current, next, ancestors);

                if (--incomingEdgesPerNode[next] == 0) {
                    queue.push(next);
                }
            }
        }
        return createListAllAncestorsPerNode(ancestors);
    }

    vector<int> createArrayIncomingEdgesPerNode(span<const vector<int>> edges) const {
        vector<int> incomingEdgesPerNode(numberOfNodes);
        for (const auto& edge : edges) {
            int to = edge[1];
            ++incomingEdgesPerNode[to];
        }
        return incomingEdgesPerNode;
    }

    vector<vector<int>> createDirectedAcyclicGraph(span<const vector<int>> edges) const {
        vector<vector<int>> directedAcyclicGraph(numberOfNodes);

        for (const auto& edge : edges) {
            int from = edge[0];
            int to = edge[1];
            directedAcyclicGraph[from].push_back(to);
        }
        return directedAcyclicGraph;
    }

    queue<int> createQueueNodesWithoutIncomingEdges(span<const int> incomingEdgesPerNode) const {
        queue<int> queue;
        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            if (incomingEdgesPerNode[nodeID] == 0) {
                queue.push(nodeID);
            }
        }
        return queue;
    }

    vector<vector<int>> createListAllAncestorsPerNode(span<const bitset<RANGE_OF_NODES[1] + 1>> ancestors) const {
        vector<vector<int>> allAncestorsPerNode(numberOfNodes);

        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            vector<int> ancestorsForCurrentNode;

            for (int candidateAncestorID = 0; candidateAncestorID < numberOfNodes; ++candidateAncestorID) {
                if (ancestors[nodeID][candidateAncestorID]) {
                    ancestorsForCurrentNode.push_back(candidateAncestorID);
                }
            }
            allAncestorsPerNode[nodeID] = ancestorsForCurrentNode;
        }
        return allAncestorsPerNode;
    }

    void updateAncestors(int current, int next, span<bitset<RANGE_OF_NODES[1] + 1>> ancestors) const {
        ancestors[next] |= ancestors[current];
        ancestors[next].set(current);
    }
};
